x-pack/platform/plugins/shared/streams/server/routes/internal/streams/systems/route.ts

```diff
@@ -0,0 +1,261 @@
+
+    const { name, systemName } = params.path;
+
+    const { write } = await checkAccess({ name, scopedClusterClient });
+
+    if (!write) {
+      throw new SecurityError(`Cannot delete system for stream ${name}, insufficient privileges`);
+    }
+
+    return await systemClient.deleteSystem(name, systemName);
+  },
+});
+
+export const updateSystemRoute = createServerRoute({
+  endpoint: 'PUT /internal/streams/{name}/systems/{systemName}',
+  options: {
+    access: 'internal',
+    summary: 'Updates a system for a stream',
+    description: 'Updates the specified system',
+  },
+  security: {
+    authz: {
+      requiredPrivileges: [STREAMS_API_PRIVILEGES.manage],
+    },
+  },
+  params: z.object({
+    path: z.object({ name: z.string(), systemName: z.string() }),
+    body: z.object({
+      description: z.string(),
+      filter: conditionSchema,
+    }),
+  }),
+  handler: async ({
+    params,
+    request,
+    getScopedClients,
+    server,
+  }): Promise<StorageClientIndexResponse> => {
+    const { systemClient, scopedClusterClient, licensing, uiSettingsClient } =
+      await getScopedClients({
+        request,
+      });
+
+    await assertSignificantEventsAccess({ server, licensing, uiSettingsClient });
+
+    const {
+      path: { name, systemName },
+      body: { description, filter },
+    } = params;
+
+    const { write } = await checkAccess({ name, scopedClusterClient });
+
+    if (!write) {
+      throw new SecurityError(`Cannot update systems for stream ${name}, insufficient privileges`);
+    }
+
+    return await systemClient.updateSystem(name, {
+      name: systemName,
+      description,
+      filter,
+    });
+  },
+});
+
+export const listSystemsRoute = createServerRoute({
+  endpoint: 'GET /internal/streams/{name}/systems',
+  options: {
+    access: 'internal',
+    summary: 'Lists all systems for a stream',
+    description: 'Fetches all systems for the specified stream',
+  },
+  security: {
+    authz: {
+      requiredPrivileges: [STREAMS_API_PRIVILEGES.read],
+    },
+  },
+  params: z.object({
+    path: z.object({ name: z.string() }),
+  }),
+  handler: async ({
+    params,
+    request,
+    getScopedClients,
+    server,
+  }): Promise<{ systems: System[] }> => {
+    const { systemClient, scopedClusterClient, licensing, uiSettingsClient } =
+      await getScopedClients({
+        request,
+      });
+
+    await assertSignificantEventsAccess({ server, licensing, uiSettingsClient });
+
+    const { name } = params.path;
+
+    const { read } = await checkAccess({ name, scopedClusterClient });
+
+    if (!read) {
+      throw new SecurityError(`Cannot read stream ${name}, insufficient privileges`);
+    }
+
+    const { hits: systems } = await systemClient.getSystems(name);
+
+    return {
+      systems,
+    };
+  },
+});
+
+export const bulkSystemsRoute = createServerRoute({
+  endpoint: 'POST /internal/streams/{name}/systems/_bulk',
```

Comment:
we use `_` as a reserved character here for a special endpoint, but we also allow to call a system `_bulk`. We should probably fail if a systems name starts with `_` - for streams themselves we do this implicitly because they map to data streams

Reply:
agreed, will add